MessageBar: Fix high contrast mode issue where you cannot tell which buttons you have focus on.

cliffkoh opened the pull request:
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #9934
- [x] Include a change request file using `$ yarn change`

#### Description of changes

Fix high contrast bugs in message bar. Also made high contrast focus rects for dismiss/expand button consistent.

##### Before: 
https://fabricweb.z5.web.core.windows.net/oufr/7.16.0/#/examples/messagebar

##### After:
http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/9953/merge/fabric-website-resources/dist/index.html#/examples/messagebar

#### Focus areas to test

(optional)


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9953)

cliffkoh commented:
@msft-github-bot please merge this once @khmakoto and @ecraig12345 has graciously given their sign off

Vitalius1 commented on packages/office-ui-fabric-react/src/components/Button/MessageBarButton/MessageBarButton.styles.ts:
```diff
@@ -2,16 +2,27 @@ import { IButtonStyles } from '../Button.types';
 import { ITheme, concatStyleSets } from '../../../Styling';
 import { memoizeFunction } from '../../../Utilities';
 import { getStyles as getBaseButtonStyles } from '../BaseButton.styles';
+import { getFocusStyle } from '@uifabric/styling';
```
This import can and should be done in line 2 import.

cliffkoh replied:
Thanks for catching :) The import was added by vscode, I believe.

ecraig12345 commented on packages/office-ui-fabric-react/src/components/MessageBar/MessageBar.styles.ts:
```diff
@@ -84,21 +85,31 @@ export const getStyles = (props: IMessageBarStyleProps): IMessageBarStyles => {
+      padding: '8px 12px',
+      selectors: {
+        '& .ms-Button-icon': dismissalAndExpandIconStyle,
+        ':hover': {
```
I know this part is just moved but this selector could be combined with the next (`:hover, :active`)

cliffkoh replied:
Combining them however has the adverse effect of making it harder for concatStyleSets to kick in, therefore making it harder to customize/overwrite these selectorts

ecraig12345 replied:
Oh interesting...mergeStyles is so hard...

Vitalius1 commented:
Interesting... when I go to try out HC mode on our live website I can see what button I focus bun in a not very useful way 😄 
![messagebar](https://user-images.githubusercontent.com/29514833/61980994-2ebf4200-afad-11e9-84ca-6c5f42e32d8e.JPG)

... so I wonder why I can see a focus happening on the live website and not on the [deploy link](https://fabricweb.z5.web.core.windows.net/oufr/7.16.0/#/examples/messagebar) @cliffkoh mentioned in the issue.

cliffkoh commented:
Live website has styles from the surrounding page loaded too, which may have caused the behavior you observed. Hence I typically use aka.ms/fabricdemo.

cliffkoh commented:
BTW the "Before" link is not a deploy link but a link from aka.ms/fabricdemo

Vitalius1 commented:
Aren’t deploy links about the same as fabric demo ones?

ecraig12345 commented:
With deploy links you have the choice of the demo site or the full website.

cliffkoh commented:
@Vitalius1 depends on what you mean by same but technically they're quite different. PR deploy links includes the demo pages for other packages as well - the fabricdemo ones are limited only to the main OUFR package (or fabric-website-resources package), has a different URL, and is limited to the demo site (not the full website).

Vitalius1 commented:
Not taking the URL into discussion how different is [this link](http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/9953/merge/fabric-website-resources/dist/index.html) which is the `fabric-website-resources` version of the deploy links and [this link](https://fabricweb.z5.web.core.windows.net/oufr/7.16.0/#/examples/messagebar) which is the `aka.ms/fabricdemo` link?

cliffkoh commented:
They're the same thing (ignoring version differences). Both are fabric-website-resources.

ecraig12345 commented:
@msft-github-bot Make sure @betrue-final-final approves

betrue-final-final commented:
Why is the focus rect outside the button in Standard, and invisible in Primary? I'm looking at the Deploy link. in the checks.
![image](https://user-images.githubusercontent.com/20363037/62009467-7c09f380-b114-11e9-83ee-ac744de6abcb.png)

betrue-final-final commented:
Maybe I'm not looking at the right link?

betrue-final-final commented:
![image](https://user-images.githubusercontent.com/20363037/62009502-be333500-b114-11e9-8e88-092936db245b.png)
"Demo site" right?

cliffkoh commented:
It looks like you're looking into Button, this PR changes MessageBar only.

Specifically compare and contrast

Before:
https://fabricweb.z5.web.core.windows.net/oufr/7.16.0/#/examples/messagebar

After:
http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/9953/merge/fabric-website-resources/dist/index.html#/examples/messagebar

betrue-final-final commented:
Ok, after this meeting

betrue-final-final commented:
Why is this different than any other button in terms of focus? That cyan will not work against white no matter how thick it is. This should really be an inset black box inside just like standalone button.
![image](https://user-images.githubusercontent.com/20363037/62082056-d3898b80-b208-11e9-8b1c-aa9dd9bd2766.png)

joschect commented:
@cliffkoh Have you addressed Ben's comments? is this ready to checkin?

cliffkoh commented:
@betrue-final-final I used the hover styles I see in some other components (and used the system color Highlight) but your suggestion makes sense particularly if there is an existing buttons for High Contrast buttons to be treated this way. Happy to make the change (after hours).

cliffkoh commented:
@betrue-final-final regarding the color, shouldn't it be the System color of "ActiveBorder" rather than whatever results in a white border like "ButtonText"?

I realize as well that the existing styles do not set an inset border.. in the Button component itself, the focus border has an outline offset of about 2px.

I will make it inset as you requested, but I will use what seems to be the more semantically correct color of "ActiveBorder" instead of "ButtonText" (which now makes me think the "Highlight" color that I see used in so many other components is wrong :\)

cliffkoh commented:
Thank you @Vitalius1 and @betrue-final-final for in-depth review and valuable feedback. Hopefully this is now good to go in (and hopefully the use of "ActiveBorder" is non-contentious).

cnamal commented:
hi @cliffkoh, any updates on this PR? :)

cliffkoh commented:
I'm waiting for @betrue-final-final to approve the PR

betrue-final-final commented:
Buttons look fine:
![image](https://user-images.githubusercontent.com/20363037/62658413-98304080-b91d-11e9-93ad-0bf657fac918.png)

But the caption and expand button focus is almost invisible. I think there's a knock on effect here:
![image](https://user-images.githubusercontent.com/20363037/62658437-a7af8980-b91d-11e9-89a5-428d29757138.png)

cliffkoh commented:
@betrue-final-final what would you propose for caption and expand?

cliffkoh commented:
Ping @betrue-final-final

cnamal commented:
hey @cliffkoh, @betrue-final-final 
apologies for putting pressure, but we are starting to get close to SLA :(. Any updates on this PR?

cliffkoh commented:
@dzearing @betrue-final-final can I merge this?

cliffkoh commented:
I propose merging this right now and if there are any additional changes required, we can track them in a new issue. The change as-is does address the core issue reported and make concrete progress on the component's accessibility.

cliffkoh commented:
![image](https://user-images.githubusercontent.com/1003246/63480296-2c35f800-c446-11e9-9256-daf3a6385138.png)

Updated to use Window (same as the foreground text) so it looks like this. This only works because I also fixed (by ovewriting) the border to be inset instead of offset. Without the overwrite, the top and bottom portions of the border will not be visible.

ecraig12345 commented:
I'm technically the codeowner of MessageBar but don't have a strong opinion on this issue, so what @cliffkoh said seems fine. 

Also, the build is failing due to missing snapshot updates.
